include/clang/AST/Expr.h

```diff
@@ -4748,27 +4752,68 @@ class CountBoundsExpr : public BoundsExpr {
   }
 };
 
+class RelativeBoundsExpr : public BoundsExpr {
+private:
+  TypeSourceInfo *AlignType;
+
+public:
+  RelativeBoundsExpr(TypeSourceInfo *Align, SourceLocation StartLoc,
+                     SourceLocation RParenLoc)
+      : BoundsExpr(RelativeBoundsExprClass, Relative, StartLoc, RParenLoc),
+        AlignType(Align) {}
+
+  explicit RelativeBoundsExpr(EmptyShell Empty)
+      : BoundsExpr(RelativeBoundsExprClass, Empty), AlignType(nullptr) {}
+
+  TypeSourceInfo *getAlignTypeInfoAsWritten() const { return AlignType; }
+  void setAlignTypeInfoAsWritten(TypeSourceInfo *Ty) { AlignType = Ty; }
+  QualType getTypeAsWritten() const { return AlignType->getType(); }
+  bool hasAlignType() const { return AlignType != nullptr; }
```

Comment:
I think it is expensive to use the TypeSourceInfo to compute the type.   All expressions store their QualType, so you can just call getType() to get the type of expression.  This includes expressions that have TypeSourceInfo.

Could you add a new field that stores that QualType, as well as getter and setter properties for that?